### PR TITLE
feat(sandbox): add Aliyun AgentRun sandbox client

### PR DESCRIPTION
**Pull request by iridescentWen**
### Summary

Adds an Aliyun (AgentRun) sandbox backend for the Agents SDK sandbox extension,
built on the async `agentrun-sdk` (`agentrun-sdk==0.0.35`, new optional `aliyun`
extra). It mirrors the existing Vercel backend lifecycle (`create` / `resume` /
`delete` / `SandboxSession` context manager) so callers can swap backends without
code changes.

Highlights:
- Async AgentRun API throughout (`create_sandbox_async`, `cmd_async`,
  `download_async`/`upload_async`, `check_health_async`, `delete_async`,
  `get_sandbox_async`) — no blocking `to_thread` wrappers.
- Production hardening: gateway HTTP 502 retry with backoff, the 30s
  `/processes/cmd` hard cap, and a `resolve-workspace-path` empty-stdout retry.
- `running()` uses the cheap health probe; adds `append()`, a 100MB upload guard,
  and OSS / NAS / PolarFs mount configs passed straight through to
  `create_sandbox_async`.
- `resume()` reattaches an existing sandbox via `get_sandbox_async` and falls back
  to a fresh sandbox + snapshot rehydration when it is gone.
- Credentials are supplied through the public `AliyunSandboxClientOptions`
  (or `AliyunSandboxClient` defaults) and resolved into an agentrun `Config`;
  serialized session state never carries secrets.
- Exposed under `agents.extensions.sandbox` behind the optional `aliyun` extra
  (lazy import guarded), with an example runner and reference docs.

### Test plan

- New unit suite `tests/extensions/sandbox/test_aliyun.py` (72 cases) mocks the
  `agentrun-sdk` via `sys.modules` injection, so it needs no real package and
  never touches Alibaba Cloud. Covers user-arg rejection, exec normalization,
  502/timeout handling, file IO, `append`, tar validation, health-based
  `running()`, shutdown/delete, workspace `persist`/`hydrate`, credential
  threading, resume reattach + fallback, and env injection. Backend module
  coverage is 92%.
- `make format`, `make lint`, `make typecheck` (mypy + pyright, 0 errors), and
  `make tests` (4920 passed, 1 skipped) all pass locally.

### Issue number

N/A

### Checks

- [x] I've added new tests, if relevant
- [x] I've run `.agents/skills/code-change-verification/scripts/run.sh`
- [x] I've confirmed all verification steps pass
- [ ] If using Codex, I've run `/review` before submitting this PR

### CODE REVIEW

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 1]
```diff
@@ -0,0 +1,1045 @@
...
+                path=posix_path_as_path(root),
+                context={
+                    "exit_code": exit_code,
+                    "stdout": stdout,
+                    "stderr": stderr,
+                },
+            )
+
+        if self.state.env:
+            await self._inject_user_env(self.state.env)
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Good catch — fixed in `23c95aec`. `_prepare_backend_workspace` now resolves `self.state.manifest.environment.resolve()` and merges it with the option-provided `state.env` (manifest wins on key collisions, `None` values dropped) before `_inject_user_env`, matching the other backends. Added test `test_prepare_backend_workspace_merges_manifest_env`.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 2]
```diff
@@ -0,0 +1,1045 @@
...
+        except Exception as exc:
+            raise WorkspaceStartError(
+                path=posix_path_as_path(coerce_posix_path(self.state.manifest.root)),
+                cause=exc,
+            ) from exc
+
+        sandbox = CodeInterpreterSandbox.model_validate(base_sandbox.model_dump(by_alias=False))
+        sandbox._config = self._config
+
+        self._sandbox = sandbox
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py:423]: Fixed in `23c95aec`. `_ensure_sandbox` now calls a new `_wait_until_healthy` right after `create_sandbox_async`, polling `check_health_async` ~once/sec until the sandbox reports healthy (up to `DEFAULT_ALIYUN_WAIT_FOR_RUNNING_TIMEOUT_S`, which is now actually wired up), mirroring the SDK's `CodeInterpreterSandbox.__enter__`. On timeout it raises `WorkspaceStartError` instead of letting the first `mkdir` hit a still-booting sandbox. Added tests `test_wait_until_healthy_polls_until_ready`, `test_wait_until_healthy_timeout_raises`, and `test_client_create_waits_for_health_timeout`.

**Comment by iridescentWen** [comment]
@codex review

Addressed both P2 findings in `23c95aec`:
- Manifest env: `_prepare_backend_workspace` now merges `manifest.environment.resolve()` with `state.env` before injection.
- Health wait: `_ensure_sandbox` polls `check_health_async` until the sandbox is ready (up to `DEFAULT_ALIYUN_WAIT_FOR_RUNNING_TIMEOUT_S`) after create, raising `WorkspaceStartError` on timeout.

New tests added and `make format/lint/typecheck/tests` all pass.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 3]
```diff
@@ -0,0 +1,1103 @@
...
+
+    async def create(
+        self,
+        *,
+        snapshot: SnapshotSpec | SnapshotBase | None = None,
+        manifest: Manifest | None = None,
+        options: AliyunSandboxClientOptions,
+    ) -> SandboxSession:
+        resolved_manifest = _resolve_manifest_root(manifest)
+        region = options.region or self._region or DEFAULT_ALIYUN_REGION
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Fixed in `dab23951`. `AliyunSandboxClientOptions.region` is now `str | None = None`, and `create` resolves it via `_resolve_credential(options.region, self._region) or DEFAULT_ALIYUN_REGION` — same `is not None` precedence as the credential fallback — so a client-level `region` now takes effect when options leaves it unset. Added `test_client_create_honors_client_region_default` and `test_client_create_options_region_overrides_client`.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 4]
```diff
@@ -0,0 +1,1103 @@
...
+        """Persist `env` into the sandbox shell so that `sh -lc` execs pick it up.
+
+        Writes `~/.openai_agents_env` with `export` lines and sources it from
+        `~/.profile` via a marker section so we don't append duplicates on
+        repeated calls.
+        """
+        sandbox = await self._ensure_sandbox()
+        env_lines = "\n".join(
+            f"export {k}={shlex.quote(v)}" for k, v in env.items() if v is not None
+        )
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Fixed in `dab23951`. `_inject_user_env` now validates each key against `^[A-Za-z_][A-Za-z0-9_]*$` before writing the `export` line; non-identifier names (e.g. `X; curl ... #`) are skipped with a warning rather than emitted raw into the sourced file. Added `test_inject_user_env_skips_unsafe_names` and `test_inject_user_env_all_names_unsafe_is_noop`. (This one was exposed by the manifest-env fix above — good catch.)

**Comment by iridescentWen** [comment]
@codex review

Addressed both P2 findings from the last pass in `dab23951`:
- Region default: `options.region` is now optional (`None`) and resolved with the same `is not None` precedence as credentials, so `AliyunSandboxClient(region=...)` takes effect.
- Env name safety: `_inject_user_env` now rejects non-identifier variable names before writing shell exports.

New tests added; `make format/lint/typecheck/tests` all pass.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 5]
```diff
@@ -0,0 +1,1116 @@
...
+                logger.warning("[AliyunSandboxSession] skipping env var with unsafe name %r", key)
+                continue
+            export_lines.append(f"export {key}={shlex.quote(value)}")
+        env_lines = "\n".join(export_lines)
+        if not env_lines:
+            return
+        script = (
+            f"cat > {_AGENTS_ENV_FILE} <<'OPENAI_AGENTS_ENV_EOF'\n"
+            f"{env_lines}\n"
+            "OPENAI_AGENTS_ENV_EOF\n"
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Fixed in `dc63eaa5`. `_inject_user_env` now registers the secret-bearing `.openai_agents_env` file (which lives at `$HOME`, i.e. the default workspace root) as a persist-workspace skip path via `register_persist_workspace_skip_path`, so `persist_workspace()` adds `--exclude=./.openai_agents_env` and the secrets never land in snapshots or resumed workspaces. `.profile` only gets a non-secret `source` line, so it may persist. Added `test_inject_user_env_excluded_from_snapshot`.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 6]
```diff
@@ -0,0 +1,1116 @@
...
+            timeout_int = ALIYUN_PROCESS_CMD_HARD_CAP_S
+        else:
+            timeout_int = max(1, min(int(timeout_s), ALIYUN_PROCESS_CMD_HARD_CAP_S))
+
+        logger.info(
+            "AliyunSandboxSession._run_command sandbox_id=%s cwd=%s timeout=%ss cmd=%r",
+            self.state.sandbox_id,
+            effective_cwd,
+            timeout_int,
+            shell_cmd[:500],
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Fixed in `dc63eaa5`. `_run_command` gained a `redact_command` flag; `_inject_user_env` calls it with `redact_command=True` so the INFO log records `cmd='<redacted>'` instead of the `export KEY=<secret>` payload. The command still executes normally against the sandbox. Added `test_inject_user_env_does_not_log_secret`.

**Comment by iridescentWen** [comment]
@codex review

Addressed both P1 secret-leak findings in `dc63eaa5`:
- Snapshot leak: the `.openai_agents_env` file is now registered as a persist-workspace skip path, so `persist_workspace()` excludes it from snapshots.
- Log leak: `_run_command` gained a `redact_command` flag; env injection uses it so `export KEY=<secret>` no longer appears in logs.

New tests added; `make format/lint/typecheck/tests` all pass.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 7]
```diff
@@ -0,0 +1,1134 @@
...
+            export_lines.append(f"export {key}={shlex.quote(value)}")
+        env_lines = "\n".join(export_lines)
+        if not env_lines:
+            return
+        # The env file holds secret values and lives at `$HOME/.openai_agents_env`,
+        # which is inside the default workspace root. Exclude it from workspace
+        # snapshots so secrets never land in persisted/resumed workspaces. (`.profile`
+        # only gets a `source` line + marker, no secrets, so it may persist.)
+        try:
+            self.register_persist_workspace_skip_path(_AGENTS_ENV_REL_PATH)
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Fixed in `7e64cf03`. The skip path is now computed relative to the actual `manifest.root`: `_inject_user_env` takes the env file's absolute path (`$HOME/.openai_agents_env`, where $HOME is AgentRun's `/home/user`) and registers `env_file_abs.relative_to(manifest.root)` — so root `/home` excludes `user/.openai_agents_env`, root `/` excludes `home/user/.openai_agents_env`, and a root that doesn't contain $HOME registers nothing (no leak, no error). Added `test_inject_user_env_skip_path_relative_to_ancestor_root` and `test_inject_user_env_skip_noop_when_env_outside_root`.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 8]
```diff
@@ -0,0 +1,1134 @@
...
+        if not normalized:
+            return ExecResult(stdout=b"", stderr=b"", exit_code=0)
+
+        try:
+            exit_code, stdout, stderr = await asyncio.wait_for(
+                self._run_command(
+                    sandbox,
+                    normalized[0],
+                    normalized[1:],
+                    cwd=self.state.manifest.root,
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py:644]: Fixed in `7e64cf03`. `_exec_internal` now passes `cwd=self.state.default_cwd or self.state.manifest.root`, so user execs honor a configured `default_cwd` and fall back to the workspace root only when it's unset. Added `test_exec_uses_default_cwd` and `test_exec_defaults_to_manifest_root_without_default_cwd`.

**Comment by iridescentWen** [comment]
@codex review

Addressed both findings from the last pass in `7e64cf03`:
- Snapshot skip (P1): the exclusion is now computed relative to the actual `manifest.root`, covering roots that are ancestors of `$HOME` (e.g. `/home`, `/`).
- default_cwd (P2): `_exec_internal` now honors a configured `default_cwd` for user execs.

New tests added; `make format/lint/typecheck/tests` all pass.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 9]
```diff
@@ -0,0 +1,1144 @@
...
+        if not normalized:
+            return ExecResult(stdout=b"", stderr=b"", exit_code=0)
+
+        try:
+            exit_code, stdout, stderr = await asyncio.wait_for(
+                self._run_command(
+                    sandbox,
+                    normalized[0],
+                    normalized[1:],
+                    cwd=self.state.default_cwd or self.state.manifest.root,
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Fixed in `11b8cbae`. The prior pass applied `default_cwd` inside `_exec_internal`, but every internal operation (runtime-helper install, workspace setup, `persist_workspace`/`hydrate_workspace`, filesystem ops) funnels through that same path — so it broke far more than persist. Since the base session offers no user-vs-internal exec distinction, and `default_cwd` was already vestigial (never applied to execs originally, unused in examples, no other backend has it), I removed the `default_cwd` option entirely: all execs now run from the workspace root. This resolves this finding and the previous 'honor default_cwd' one together without a half-working option. Updated the exec test to `test_exec_runs_in_manifest_root`.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 10]
```diff
@@ -0,0 +1,1144 @@
...
+        logger.info(
+            "AliyunSandboxSession created sandbox: template=%s sandbox_id=%s",
+            template_name,
+            sandbox.sandbox_id,
+        )
+        # A freshly created sandbox may still be booting; the SDK's own
+        # CodeInterpreterSandbox context manager polls health for this reason.
+        # We bypass that context manager, so wait explicitly before the first
+        # exec (e.g. the workspace mkdir) can hit a not-yet-ready sandbox.
+        await self._wait_until_healthy(sandbox)
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Fixed in `11b8cbae`. `_ensure_sandbox` now wraps the health wait: if `_wait_until_healthy` times out or raises, it calls `_stop_attached_sandbox()` (best-effort `delete_sandbox_async`) before re-raising, so a sandbox that never becomes healthy is deleted instead of leaking until its idle timeout. Extended `test_client_create_waits_for_health_timeout` to assert the sandbox is deleted.

**Comment by iridescentWen** [comment]
@codex review

Addressed both findings from the last pass in `11b8cbae`:
- Internal exec cwd (P1): removed the vestigial `default_cwd` option; all execs now run from the workspace root, so internal ops (persist/hydrate, runtime helpers, workspace setup) are no longer redirected. This also resolves the earlier 'honor default_cwd' P2, since a user/internal exec distinction isn't available and the option was never functional.
- Health-timeout cleanup (P2): `_ensure_sandbox` now deletes the sandbox if it never becomes healthy, instead of leaking it.

New/updated tests; `make format/lint/typecheck/tests` all pass.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 11]
```diff
@@ -0,0 +1,1146 @@
...
+            base_sandbox = await client.get_sandbox_async(sandbox_id=sandbox_id)
+        except Exception as exc:
+            if _is_not_found_error(exc):
+                return None, None
+            raise
+
+        sandbox = CodeInterpreterSandbox.model_validate(base_sandbox.model_dump(by_alias=False))
+        sandbox._config = config
+        logger.info("AliyunSandboxClient reattached sandbox: sandbox_id=%s", sandbox_id)
+        return sandbox, client
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py:1116]: Fixed in `cc22ab9b`. `_reattach_sandbox` now health-checks the reattached sandbox: after `get_sandbox_async`, it calls `check_health_async` and returns `(None, None)` if the instance isn't ready (or the probe fails), so `resume` falls through to provisioning a fresh sandbox + snapshot restore instead of reusing an expired/stopped handle. Added `test_client_resume_reattach_unhealthy_falls_back`.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 12]
```diff
@@ -0,0 +1,1146 @@
...
+
+        raw_bytes = bytes(raw)
+        root = self._workspace_root_path()
+        sandbox = await self._ensure_sandbox()
+        archive_path = posix_path_as_path(
+            coerce_posix_path(f"/tmp/openai-agents-aliyun-{self.state.session_id.hex}.tar")
+        )
+        tar_command = ("tar", "xf", archive_path.as_posix(), "-C", root.as_posix())
+        try:
+            self._validate_tar_bytes(raw_bytes)
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py:899]: Fixed in `cc22ab9b`. `_validate_tar_bytes` now passes `allow_external_symlink_targets=False` to `validate_tarfile`, matching the other hosted backends, so an archive with a symlink pointing outside `manifest.root` is rejected before the remote `tar xf`. Added `test_validate_tar_bytes_external_symlink_rejected` (verified it passes with the old default `True` and is rejected with `False`).

**Comment by iridescentWen** [comment]
@codex review

Addressed both findings from the last pass in `cc22ab9b`:
- Reattach health (P1): `_reattach_sandbox` now health-checks the sandbox returned by `get_sandbox_async` and falls back to a fresh sandbox when it's not ready.
- External symlinks (P2): `_validate_tar_bytes` now passes `allow_external_symlink_targets=False`, matching the other hosted backends, so external-symlink archives are rejected before remote extraction.

New tests added; `make format/lint/typecheck/tests` all pass.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 13]
```diff
@@ -0,0 +1,1164 @@
...
+            try:
+                self.register_persist_workspace_skip_path(skip_rel.as_posix())
+            except Exception:  # noqa: BLE001 — skip registration is best-effort.
+                logger.exception(
+                    "[AliyunSandboxSession] failed to register env file as a snapshot skip path"
+                )
+        script = (
+            f"cat > {_AGENTS_ENV_FILE} <<'OPENAI_AGENTS_ENV_EOF'\n"
+            f"{env_lines}\n"
+            "OPENAI_AGENTS_ENV_EOF\n"
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Fixed in `5eb654e5`. `_inject_user_env` no longer uses a here-doc: the env file is now transported as `printf %s '<base64>' | base64 -d > $FILE`. base64 output contains no shell-significant characters (and no newlines), so an env value can't smuggle a here-doc terminator or any shell command into the setup script — the whole payload stays opaque until it's decoded into the (chmod 600) file. Keys are still validated, values are still `shlex.quote`d for when the file is later sourced, and the command remains `redact_command=True`. Added `test_inject_user_env_value_with_heredoc_delimiter_is_safe` (asserts the injected `OPENAI_AGENTS_ENV_EOF` / `touch /pwned` never appear as literal shell, but are preserved verbatim in the decoded file).

**Comment by iridescentWen** [comment]
@codex review

Addressed the here-doc injection finding in `5eb654e5`: `_inject_user_env` now transports the env file as base64 (`printf %s '<b64>' | base64 -d > $FILE`) instead of a here-doc, so env values can't terminate a delimiter early or inject commands during start. New test added; `make format/lint/typecheck/tests` all pass.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 14]
```diff
@@ -0,0 +1,1170 @@
...
+        argv = [command, *args]
+        shell_cmd = shlex.join(argv)
+        effective_cwd = cwd or self.state.manifest.root
+        # AgentRun's /processes/cmd is hard-capped at 30s by the data-plane gateway.
+        # Always pass an int so the SDK forwards it as-is, and clamp so callers see a
+        # clean ExecTimeoutError on our side instead of an opaque gateway 504.
+        if timeout_s is None:
+            timeout_int = ALIYUN_PROCESS_CMD_HARD_CAP_S
+        else:
+            timeout_int = max(1, min(int(timeout_s), ALIYUN_PROCESS_CMD_HARD_CAP_S))
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Fixed in `76c1d207`. Confirmed against the SDK: `ProcessOperations.cmd_async(..., timeout=30)` just forwards the timeout (default 30, no enforced ceiling) — the 30s cap was a deployment-specific assumption. `_run_command` no longer clamps: it forwards the caller's timeout as-is (defaulting to 30 when unset), matching the outer `wait_for`, so a command the caller allowed more time for isn't killed early. If a specific gateway enforces a shorter limit it surfaces as a provider error rather than a silent cap. Renamed the misleading `ALIYUN_PROCESS_CMD_HARD_CAP_S` → `DEFAULT_ALIYUN_EXEC_TIMEOUT_S`; updated tests (`test_run_command_forwards_caller_timeout`, `test_run_command_defaults_timeout_when_unset`).

File: examples/sandbox/extensions/README.md [hunk 15]
```diff
@@ -350,4 +350,54 @@ The runner also includes standalone demos for individual features. Pass
...
+
+- <https://www.alibabacloud.com/help/en/agentrun>
+- <https://help.aliyun.com/document_detail/2839106.html>
+
+Export the required environment variables:
+
+```bash
+export OPENAI_API_KEY=...
+export ALIBABA_CLOUD_ACCESS_KEY_ID=...
+export ALIBABA_CLOUD_ACCESS_KEY_SECRET=...
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply examples/sandbox/extensions/README.md:379]: Fixed in `76c1d207`. Confirmed the SDK's `Config.get_data_endpoint()` builds `https://{account_id}.agentrun-data.{region}.aliyuncs.com` and `get_account_id()` raises `account id is not set` when it's missing and no data endpoint is set. Updated the README setup to export `AGENTRUN_ACCOUNT_ID` (or `FC_ACCOUNT_ID` / `AGENTRUN_DATA_ENDPOINT`, or `--account-id`), and `_require_aliyun_credentials` in the runner now enforces an account id (or data endpoint) up front with a clear message instead of letting users fail later at sandbox creation.

**Comment by iridescentWen** [comment]
@codex review

Addressed both findings from the last pass in `76c1d207`:
- Exec timeout (P2): `_run_command` no longer clamps to 30s; it forwards the caller's timeout as-is (SDK's `cmd_async` doesn't enforce a ceiling), consistent with the outer `wait_for`.
- Account id (P2): README now documents `AGENTRUN_ACCOUNT_ID` (required for the data-plane endpoint) and the runner enforces it up front.

New/updated tests; `make format/lint/typecheck/tests` all pass.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 16]
```diff
@@ -0,0 +1,1172 @@
...
+
+    async def resume(self, state: SandboxSessionState) -> SandboxSession:
+        if not isinstance(state, AliyunSandboxSessionState):
+            raise TypeError("AliyunSandboxClient.resume expects an AliyunSandboxSessionState")
+
+        config = _build_config(
+            access_key_id=self._access_key_id,
+            access_key_secret=self._access_key_secret,
+            account_id=self._account_id,
+            api_key=self._api_key,
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py:1055]: Fixed in `edadaa86`. `create()` now caches the resolved option-over-client credentials back onto the client (`_access_key_id` etc.), so a create()-then-resume() flow on the same client keeps working when credentials were supplied only via `AliyunSandboxClientOptions`. Serialized state still never carries secrets — resume rebuilds Config from the client fields, which are now populated. Added `test_client_resume_uses_option_credentials_from_create`.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 17]
```diff
@@ -0,0 +1,1172 @@
...
+        effective_cwd = cwd or self.state.manifest.root
+        # Forward the caller's timeout as-is (defaulting to the SDK default) rather
+        # than silently shortening it: the outer `wait_for` in `_exec_internal` uses
+        # the same value, so a longer command a caller allowed time for isn't killed
+        # early. If a specific AgentRun gateway enforces a shorter ceiling, that
+        # surfaces as a provider error instead of a silent cap.
+        if timeout_s is None:
+            timeout_int = DEFAULT_ALIYUN_EXEC_TIMEOUT_S
+        else:
+            timeout_int = max(1, int(timeout_s))
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Fixed in `edadaa86`. `_run_command` now uses `math.ceil(timeout_s)` instead of `int(...)`, so a fractional budget like 1.9 forwards as 2 (not 1) and the provider can't kill the command before the outer `wait_for`. Added `test_run_command_rounds_up_fractional_timeout`.

File: examples/sandbox/extensions/aliyun_runner.py [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+    access_key_id: str | None,
+    access_key_secret: str | None,
+    account_id: str | None,
+    api_key: str | None,
+) -> None:
+    env = os.environ
+    has_keys = bool(access_key_id and access_key_secret) or bool(
+        env.get("ALIBABA_CLOUD_ACCESS_KEY_ID") and env.get("ALIBABA_CLOUD_ACCESS_KEY_SECRET")
+    )
+    if not (has_keys or (account_id and api_key)):
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply examples/sandbox/extensions/aliyun_runner.py]: Fixed in `edadaa86`. The runner preflight now only enforces the account-id requirement (the one hard prerequisite for the data-plane endpoint) and leaves the access credentials to the SDK, which resolves them from CLI flags, env vars, an Alibaba Cloud config file, or a RAM role — matching what the README describes. It no longer exits early when a supported SDK credential provider is in use with `AGENTRUN_ACCOUNT_ID` set.

**Comment by iridescentWen** [comment]
@codex review

Addressed all three findings from the last pass in `edadaa86`:
- Resume credentials (P2): create() caches resolved option credentials on the client so resume() works with option-only creds (state still carries no secrets).
- Fractional timeout (P2): _run_command rounds the timeout up (math.ceil) instead of truncating.
- Runner preflight (P3): now only requires the account id and lets the SDK resolve access credentials from its providers.

New tests added; `make format/lint/typecheck/tests` all pass.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 19]
```diff
@@ -0,0 +1,1186 @@
...
+        # Resolve option-over-client credentials, and cache the result back onto the
+        # client. `resume()` only receives serialized state (which never carries
+        # secrets), so it rebuilds Config from these client fields — caching here lets
+        # a create()-then-resume() flow on the same client work even when credentials
+        # were supplied only via options.
+        self._access_key_id = self._resolve_credential(options.access_key_id, self._access_key_id)
+        self._access_key_secret = self._resolve_credential(
+            options.access_key_secret, self._access_key_secret
+        )
+        self._account_id = self._resolve_credential(options.account_id, self._account_id)
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Fixed in `2c9329ad` — and this reverts the client-caching from my previous reply, which you correctly flagged as unsafe. `create()` no longer writes resolved credentials back onto the client; the option-over-client resolution is used only to build THIS session's Config (scoped to the session). So a client reused across tenants via per-call options no longer has its shared credentials mutated, and a later resume() of an earlier session can't run under the wrong account. Trade-off: resume() relies on client-level credentials (or the SDK's env/credential providers), since per-call option secrets can't be safely persisted (state carries no secrets by design). Added `test_create_does_not_mutate_client_credentials`.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 20]
```diff
@@ -0,0 +1,1186 @@
...
+        is_resolve_helper = "resolve-workspace-path" in command
+        max_attempts = 1 + ALIYUN_GATEWAY_502_RETRIES
+
+        exit_code, stdout, stderr = 0, "", ""
+        for attempt in range(1, max_attempts + 1):
+            try:
+                result = await sandbox.process.cmd_async(
+                    command=shell_cmd,
+                    cwd=effective_cwd,
+                    timeout=timeout_int,
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py:535]: Fixed in `2c9329ad`. Once the env file has been written, `_run_command` prepends `. $HOME/.openai_agents_env 2>/dev/null;` to the command it sends to AgentRun (whose cmd API has no env parameter). That gives shell=False direct execs — and SDK helpers that use shell=False — the configured environment too, not just the `sh -lc` login-shell path that reads `.profile`. Sessions without configured env are unaffected (no prefix). Added `test_exec_shell_false_sources_env_after_injection` and `test_exec_without_env_has_no_source_prefix`.

**Comment by iridescentWen** [comment]
@codex review

Addressed both findings from the last pass in `2c9329ad`:
- Credential scoping (P2): reverted the client-credential caching (it was unsafe for multi-tenant reuse); option credentials are now scoped to the session's Config only, and create() no longer mutates shared client state.
- Env for direct execs (P2): once the env file exists, _run_command sources it inline so shell=False execs also get the configured environment.

New tests added; `make format/lint/typecheck/tests` all pass.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 21]
```diff
@@ -0,0 +1,1196 @@
...
+        # per-session `env` from client options, mirroring the other backends. The
+        # manifest takes precedence on key collisions; `None` values are dropped.
+        manifest_env = cast(dict[str, str | None], await self.state.manifest.environment.resolve())
+        env = {
+            key: value
+            for key, value in {**(self.state.env or {}), **manifest_env}.items()
+            if value is not None
+        }
+        if env:
+            await self._inject_user_env(env)
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Fixed in `$(git rev-parse --short HEAD)`. Env injection moved from `_prepare_backend_workspace()` to a new `_after_start()` override (extracted into `_apply_configured_env()`). `BaseSandboxSession.start()` runs `_after_start()` after `_start_workspace()`, i.e. after any snapshot restore — so the env file (which is excluded from snapshots) is written after the restore instead of being wiped by it. This covers both fresh create and resume. Tests renamed to `test_after_start_injects_env` / `test_after_start_merges_manifest_env`.

[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: (correction: the commit is `e1aad243` — my previous reply mis-rendered the SHA.)

**Comment by iridescentWen** [comment]
@codex review

Addressed the snapshot/env-ordering finding in `e1aad243`: env injection moved to a `_after_start()` override (runs after `_start_workspace()`/snapshot restore), so the env file — excluded from snapshots — is written after the restore instead of being wiped by it, on both fresh create and resume. New tests; `make format/lint/typecheck/tests` all pass.

File: src/agents/extensions/sandbox/aliyun/sandbox.py [hunk 22]
```diff
@@ -0,0 +1,1207 @@
...
+            )
+
+    async def _after_start(self) -> None:
+        await super()._after_start()
+        # Inject env here — after `_start_workspace()` (and any snapshot restore on
+        # resume) — rather than in `_prepare_backend_workspace()`. The env file is
+        # deliberately excluded from snapshots, so injecting it before the restore
+        # would let the restore wipe it and leave resumed execs sourcing a missing
+        # file. Running after start covers both fresh create and resume.
+        await self._apply_configured_env()
```
[the remark this answers is not in the document]
Comment (iridescentWen) [reply src/agents/extensions/sandbox/aliyun/sandbox.py]: Fixed in `d2de6c01` by replacing the env mechanism entirely, per maintainer guidance to follow how the other backends handle env. Instead of writing a `~/.openai_agents_env` file and sourcing it from `.profile`, the resolved environment (options + manifest, manifest wins, unsafe names dropped, values `shlex.quote`d) is now inlined as a `KEY=val ...` prefix on every command — the same per-exec model as e2b/daytona/runloop, adapted to AgentRun (whose cmd API has no env parameter). Because env travels inline with each command and touches no workspace file, this resolves this finding and the whole class it belonged to at once: it's available during manifest/mount application, survives snapshot restore, needs no login shell (works for shell=False), and never lands in snapshots or logs. Net: the here-doc/base64/`.profile`/snapshot-skip machinery (and its edge cases) is gone. New env tests cover the prefix, manifest/option merge, unsafe-name skipping, metachar quoting, shell=False, and log redaction.

**Comment by iridescentWen** [comment]
@codex review

Reworked env handling in `d2de6c01` per maintainer guidance to mirror the other backends. AgentRun's cmd API has no env parameter, so — like e2b/daytona/runloop inject env per exec — the resolved environment is now inlined as a `KEY=val ...` prefix on every command instead of a `.profile`-sourced file. This removes the file/here-doc/snapshot machinery and resolves the whole env edge-case class at once (available during manifest application, survives snapshot restore, works for shell=False, never in snapshots or logs). New tests added; `make format/lint/typecheck/tests` all pass.

**Comment by iridescentWen** [comment]
Closing in favor of a fresh PR to reset the review thread (this one accumulated many review rounds). The same, finalized commit will be re-submitted as a new PR.
